Develop into Prod

IlyaM70 opened the pull request.

Commit:
Add .gitattributes, .gitignore, and README.md.

Commit:
Add project files.

Commit:
Unit Test project Added

Commit:
Service layer added

Commit:
added authorization & authentication
added roles
restritced access to admin & manager sections

Commit:
products added

Commit:
Added main page filtration
Added Dbinitialiser class

Commit:
completed CRUD for restaurant types

Commit:
bug fixes

Commit:
Merge branch 'empty' into review

Commit:
Meal Service created
Add Meal View Added

Commit:
Type of restaurant controller added

Commit:
AddRestaurant moved to Restaurant contoller

Commit:
Meal controller added

Commit:
Add meal & Edit meal

Commit:
delete meal

Commit:
delete meal deletes image - done
change meal image - done
edit meal bug - fixed

Commit:
category of meal controller - done
category of meal service - done
categories in menu in view - done

Commit:
link menu and main in Restaurant view - done

Commit:
Refactoring

Commit:
added cart & orderDetail entities
added cart service

Commit:
Cart service - done

Commit:
added ShoppingCart view models
added Shopping cart controller

Commit:
plus\minus partial input added

Commit:
plus-minus input in cart - done

Commit:
plus function in cart done
cart id how assigns by session value

Commit:
CardId initialize refactored

Commit:
Cart in Restaurant view

Commit:
Mini cart (cart in restaurant view) - done

Commit:
empty mini cart - done

NikolayCheremnov commented on UnitTests/UnitTest1.cs:
```diff
@@ -0,0 +1,16 @@
+namespace UnitTests
+{
+    public class Tests
```
Этот класс надо удалить из проекта.

IlyaM70 replied:
Удален

NikolayCheremnov commented on WebFood/Controllers/AccessController.cs:
```diff
@@ -0,0 +1,116 @@
+                    ModelState.AddModelError("Email", "Уже есть пользователь с таким email");
+                };
+            }
+
```
Зачем так много пустых строк?

IlyaM70 replied:
Исправил

NikolayCheremnov commented on WebFood/Controllers/CategoryOfMealController.cs:
```diff
@@ -0,0 +1,151 @@
+
+
+        [HttpGet]
+        [Authorize(Roles = "Administrator, Manager")]
```
Может имеет смысл завести класс со статическими публичными константными полями и использовать их вместо простых констант для ролей? Типа Roles.ADMINISTRATOR или что-то вроде того.

IlyaM70 replied:
Создал класс Roles с константными полями ролей

NikolayCheremnov commented on WebFood/Controllers/CategoryOfMealController.cs:
```diff
@@ -0,0 +1,151 @@
+        }
+
+
+
```
Too many empty lines.

IlyaM70 replied:
Fixed

NikolayCheremnov commented on WebFood/Controllers/MealController.cs:
```diff
@@ -0,0 +1,190 @@
+
+                return View(viewModel);
+            }
+            else
```
В целом в подобных ситуациях else является лишним.

IlyaM70 replied:
Убрал else

NikolayCheremnov commented on WebFood/Controllers/MealController.cs:
```diff
@@ -0,0 +1,190 @@
+            {
+                return View(meal);
+            }
+            else
```
else определенно лишний, просто return оставить следом.

IlyaM70 replied:
Убрал else

NikolayCheremnov commented on WebFood/Controllers/RestaurantController.cs:
```diff
@@ -0,0 +1,246 @@
+                if (ModelState.IsValid)
+                {
+
+                    string filePath = "wwwroot/" + restaurant.Imageurl;
```
wwwroot/ хорошо бы вынести в конфиг appsettings.json и дергать оттуда

IlyaM70 replied:
Вынес в appsettings.json

NikolayCheremnov commented on WebFood/Controllers/ShoppingCartController.cs:
```diff
@@ -0,0 +1,110 @@
+            // Display the confirmation message
+            var results = new ShoppingCartAddViewModel
+            {
+                Message = "Блюдо " + meal.Name + " добавлено в корзину",
```
Давайте использовать интерполяцию строк

IlyaM70 replied:
Исправил

NikolayCheremnov commented on WebFood/Controllers/ShoppingCartController.cs:
```diff
@@ -0,0 +1,110 @@
+            // Display the confirmation message
+            var results = new ShoppingCartRemoveViewModel
+            {
+                Message = "Блюдо " + mealName + " убрано из корзины",
```
should use string interpolation

IlyaM70 replied:
Fixed

NikolayCheremnov commented on WebFood/Models/Entities/Cart.cs:
```diff
@@ -0,0 +1,21 @@
+    public class Cart
+    {
+        [Key]
+        public int RecordId { get; set; }
```
Точно такой атрибут нужен? Не [PrimaryKey] ?

NikolayCheremnov replied:
Или может ForeignKey. Для чего Key?

IlyaM70 replied:
Атрибут [Key] явно указывает Entity Framework, что это PrimaryKey, здесь я его поставил для того, чтобы EF не назначил свойство CardId как PrimaryKey. CardId в данной случае id всей корзины, а сущность представляет собой предметы в корзине

NikolayCheremnov commented on WebFood/Models/AppDbContext.cs:
```diff
@@ -0,0 +1,27 @@
+        {
+        }
+
+        public DbSet<User> Users { get; set; }
```
А в сущностях есть навигационные свойства-коллекции?

IlyaM70 replied:
Есть

NikolayCheremnov commented on WebFood/Models/Entities/Meal.cs:
```diff
@@ -0,0 +1,32 @@
+
+        [Range(1,10000,ErrorMessage ="Введите цену")]
+        public decimal Price { get; set; }
+        public DateTime CreatedDate { get; set;}
```
В БД нет проблем с DateTime?

IlyaM70 replied:
Проблем нет, по крайней мере с MySql

NikolayCheremnov commented on WebFood/Program.cs:
```diff
@@ -0,0 +1,99 @@
+if (!app.Environment.IsDevelopment())
+{
+    app.UseExceptionHandler("/Home/Error");
+    // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
```
Что значит этот коммент? Тут не будет проблем?

IlyaM70 replied:
Это сгенерированный код, пока убрал, потому что не знаю нужно ли мне использовать HSTS

NikolayCheremnov commented on WebFood/Service/CartService/CartService.cs:
```diff
@@ -0,0 +1,170 @@
+
+namespace WebFood.Service.CartService
+{
+    public class CartService : ICartService
```
Было бы круто использовать асинхронные методы в сервисах, но если везде синхронные были, то можно опустить.

IlyaM70 replied:
Изменил методы на асинхронные

NikolayCheremnov commented on WebFood/Service/MealService/DaoMeal.cs:
```diff
@@ -0,0 +1,50 @@
+
+        public void Delete(Meal meal)
+        {
+            string filePath = "wwwroot/" + meal.ImageUrl;
```
Хорошо бы вынести в конфиги приложения.

IlyaM70 replied:
Вынес

NikolayCheremnov commented on WebFood/Service/RestaurantService/DaoRestaurant.cs:
```diff
@@ -0,0 +1,92 @@
+
+            if (File.Exists(filePath))
+            {
+                File.Delete(filePath);
```
Вообще я бы посмотрел в сторону хранения картинок в БД в виде бинарного поля.

IlyaM70 replied:
Пока оставлю, если будет время до защиты, попробую сделать

NikolayCheremnov commented on WebFood/Utility/PasswordHasher.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace WebFood.Utility
+{
+    public static class PasswordHasher
```
Хорошо бы ипользовать паттерн "стратегия" для хэшера паролей.

IlyaM70 replied:
Реализовал паттерн для хэшера паролей

NikolayCheremnov commented on WebFood/wwwroot/lib/bootstrap/dist/css/bootstrap-grid.css:
Веб-либы надо в .gitignore добавить.

IlyaM70 replied:
Добавил

Commit:
UnitTest1.cs deleted
Added Roles class for roles constants
Added image files path in appsettings
Cart service methods now async
Refactoring

Commit:
Password hasher uses patter "strategy"

Commit:
removed use HTPS

Commit:
Added Order service, order controller, address and payment view

Commit:
Moq payment - done

Commit:
one order - one restaurant - done

Commit:
Restaurant Orders - done

Commit:
Profile - done
Profile info - done
Profile Edit - done
Profile Orders - done
Cancel Order - done
Repeat Order - done

Commit:
Manager not allowed to use cart

Commit:
Manager not allowed in profile

Commit:
Change db to ElephantSQL(PostgreSQL) - done

Commit:
Update gitingnore

IlyaM70 committed:
Update appsettings.json

Removed connection strings
